internal/controllers/provisioningrequest_controller.go

```diff
@@ -2558,3 +2520,46 @@ func (r *ProvisioningRequestReconciler) SetupWithManager(mgr ctrl.Manager) error
 			})).
 		Complete(r)
 }
+
+const (
+	// Policy Specific states
+	compliant    = "Compliant"
+	nonCompliant = "NonCompliant"
+	pending      = "Pending"
+)
+
+// BuildPoliciesEventHistory creates an EventHistory object from a policyList
+func BuildPoliciesEventHistory(policies *policiesv1.PolicyList) (h utils.EventHistory) {
+	for _, policy := range policies.Items {
```

Comment:
From my understanding, if a policy is `Compliant` it can be skipped.

Also, I don't think we need to build such a big structure with all the history from all the policies. I don't see why we can't call `IsTimedOut` for each policy or even `detailsPerTemplate` and if any of them has timed out, then that's the final result and we can return without processing the history of other policies.

Reply:
Yeah, I am concerned about the eventHistory as it could potentially become very long. It seems like there's room for optimization.

Reply:
Performance impact should be limited, seems like only 10 events Max are stored per policy. But I agree, it could be a concern for scaling to managing 3500 SNOs. planning to close the PR for now